msal/http_decorate.py

```diff
@@ -0,0 +1,117 @@
+from threading import Lock
+from hashlib import sha256
+
+from .individual_cache import _IndividualCache as IndividualCache
+from .individual_cache import _ExpiringMapping as ExpiringMapping
+
+
+# https://datatracker.ietf.org/doc/html/rfc8628#section-3.4
+DEVICE_AUTH_GRANT = "urn:ietf:params:oauth:grant-type:device_code"
+
+
+def _hash(raw):
+    return sha256(repr(raw).encode("utf-8")).hexdigest()
+
+
+def _handle_http_429_5xx_retry_after(result=None, **ignored):
+    assert result is not None, """
+        The signature defines it with a default value None,
+        only because the its shape is already decided by the
+        IndividualCache's.__call__().
+        In actual code path, the result parameter here won't be None.
+        """
+    response = result
+    lowercase_headers = {  # MSAL's HttpClient may not have headers
+        k.lower(): v for k, v in getattr(response, "headers", {}).items()}
+    default = 600
+    try:
+        # AAD's retry_after uses integer format only
+        # https://stackoverflow.microsoft.com/questions/264931/264932
+        retry_after = int(lowercase_headers.get("retry-after", default))
+    except ValueError:
+        retry_after = default
```

Comment:
If I'm reading this correctly, when we get a response whose Retry-After is not a number, we cache that response for 10 minutes? Probably that never happens but it would be an interesting bug if it ever did. Do you think it's reasonable to just forego the default and assume all responses are from AAD?

Reply:
The reasoning behind the current implementation was that the presence of `Retry-After` would mean the service wants the client to cool down for *some time*. When the current implementation does not (yet?) support all types of input, it seems a moderate default value would be close enough to honor the "spirit of the `Retry-After` header". It should be better of an uncaught exception, which could nullify the throttling effect.

We might, however, consider to also add support to the other time format of `Retry-After`. But, even after that, this default fallback branch would likely stay.

Reply:
UPDATE: In the latest commit, I added support to another popular date format, even though it is not (currently?) used by AAD; I also changed the default value from 600 to 60, which is hopefully less aggressive now.

Closing this conversation.